pkg/cephfs/nodeserver.go

```diff
@@ -308,6 +309,89 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 	return &csi.NodeUnstageVolumeResponse{}, nil
 }
 
+func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		err = fmt.Errorf("targetpath %v is empty", targetPath)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	volID := req.GetVolumeId()
+	/* TODO: Map the volumeID to the targetpath.
+
+	we need secret to connect to the ceph cluster to get the volumeID from volume
+	Name, however `secret` field/option is not available  in NodeGetVolumeStats spec,
+	Below issue covers this request and once its available, we can do the validation
+	as per the spec.
+	https://github.com/container-storage-interface/spec/issues/371
+
+	*/
+
+	isMnt, err := isMountPoint(targetPath)
+
+	if err != nil {
+		klog.Errorf("stat failed on path: %v, error: %v", targetPath, err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	if isMnt {
+		klog.Infof("cephfs: volume %s is already mounted to %s", volID, targetPath)
+		cephfsProvider := volume.NewMetricsStatFS(targetPath)
+		volMetrics, volMetErr := cephfsProvider.GetMetrics()
+		if volMetErr != nil {
+			return nil, status.Error(codes.Internal, volMetErr.Error())
+		}
+
+		available, ok := (*(volMetrics.Available)).AsInt64()
+		if !ok {
+			klog.Errorf("failed to fetch available bytes")
+		}
+		capacity, ok := (*(volMetrics.Capacity)).AsInt64()
+		if !ok {
+			klog.Errorf("failed to fetch capacity bytes")
+		}
+		used, ok := (*(volMetrics.Used)).AsInt64()
+		if !ok {
+			klog.Errorf("failed to fetch used bytes")
+		}
+		inodes, ok := (*(volMetrics.Inodes)).AsInt64()
+		if !ok {
+			klog.Errorf("failed to fetch available inodes")
+		}
+		inodesFree, ok := (*(volMetrics.InodesFree)).AsInt64()
+		if !ok {
+			klog.Errorf("failed to fetch free inodes")
+		}
+
+		inodesUsed, ok := (*(volMetrics.InodesUsed)).AsInt64()
+		if !ok {
+			klog.Errorf("failed to fetch used inodes")
+		}
+		return &csi.NodeGetVolumeStatsResponse{
+			Usage: []*csi.VolumeUsage{
+				{
+					Available: available,
+					Total:     capacity,
+					Used:      used,
+					Unit:      1,
+				},
+				{
+					Available: inodesFree,
+					Total:     inodes,
+					Used:      inodesUsed,
+					Unit:      2,
+				},
+			},
+		}, nil
+
+	}
+
+	klog.Infof("cephfs: successfully got stats for volume %s to %s", volID, targetPath)
```

Comment:
Reduce verbosity, move this to v(3/4)?

Reply:
IMO, this has to be logged. so keeping unchanged.

Reply:
This is outside the isMnt check as well, so the message is actually misleading. If the intent is to keep this for informational or troubleshooting purposes, to understand which `targetPath` or 'VolumeID` was responded to, that information is already present in the gRPC request that is logged.

Anyway, this may at least need moving into the if block above to log on success. Is this to be a followup?

Reply:
done.